clippy_lints/src/path_from_format.rs

```diff
@@ -0,0 +1,100 @@
+}
+
+declare_lint_pass!(PathFromFormat => [PATH_FROM_FORMAT]);
+
+impl<'tcx> LateLintPass<'tcx> for PathFromFormat {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if_chain! {
+            if let ExprKind::Call(_, args) = expr.kind;
+            if let ty = cx.typeck_results().expr_ty(expr);
+            if is_type_diagnostic_item(cx, ty, sym::PathBuf);
+            if !args.is_empty();
+            if let Some(macro_call) = root_macro_call(args[0].span);
+            if cx.tcx.item_name(macro_call.def_id) == sym::format;
+            if let Some(format_args) = FormatArgsExpn::find_nested(cx, &args[0], macro_call.expn);
+            let mut applicability = Applicability::MachineApplicable;
+            let format_args_snip = snippet_with_applicability(cx, format_args.inputs_span(), "..", &mut applicability);
+            if let Some(end_of_literal) = format_args_snip.find("\",");
+            then {
+                let (literal, vars) = format_args_snip.split_at(end_of_literal);
+                let mut literal = literal.to_string();
+                literal.remove(0);
+                let v: Vec<&str> = literal.split("{}").collect();
+                let real_vars = vars.strip_prefix("\", ").unwrap_or(vars);
+                if v.len() != 2 || real_vars.contains(',') {
+                    span_lint_and_help(
+                        cx,
+                        PATH_FROM_FORMAT,
+                        expr.span,
+                        "`format!(..)` used to form `PathBuf`",
+                        None,
+                        "consider using `.join()` to avoid the extra allocation",
+                    );
+                    return;
+                }
+                let sugg = {
+                    if v[0].is_empty() {
+                        let mut str1 = v[1].to_string();
+                        if str1.starts_with('\\') || str1.starts_with('/') {
+                            str1.remove(0);
+                        }
+                        format!("Path::new({real_vars}).join(\"{str1}\")")
+                    }
+                    else if v[1].is_empty() {
+                        let str1 = v[0].to_string();
+                        format!("Path::new(\"{str1}\").join({real_vars})")
+                    }
+                    else {
+                        let (str1, mut str2) = (v[0].to_string(), v[1].to_string());
+                        if str2.starts_with('\\') || str2.starts_with('/') {
+                            str2.remove(0);
+                        }
+                        format!("Path::new(\"{str1}\").join({real_vars}).join(\"{str2}\")")
+                    }
+                };
+                span_lint_and_sugg(
+                    cx,
+                    PATH_FROM_FORMAT,
+                    expr.span,
+                    "`format!(..)` used to form `PathBuf`",
+                    "consider using `.join()` to avoid the extra allocation",
```

Comment:
This is incorrent: `PathBuf::from(String)` is allocation-free, while `.join()` allocates a new `PathBuf` on every call, so using `.join()` will only result in more allocations, not fewer.

Reply:
Thanks! What should be put there instead? Is it just about code readability, or is it more efficient in some other way?

Reply:
The main advantage is that, `.join()` uses the platform-specific path separator (`/` on Unix and `\` on Windows), so the path is more portable. Readability could also count. I couldn't find a discussion for Rust, but here's the same discussion for Python: https://stackoverflow.com/questions/13944387/why-use-os-path-join-over-string-concatenation

By the way, it is possible to avoid the extra allocation of `.join()` by using `PathBuf::push` instead. I suggest adding this information to the known problems section in the lint description.